src/checkpoint.hh

```diff
@@ -934,6 +1019,124 @@ SUPPORTED_EXTERNAL_TYPES = {"Rope", "answer_pknot_mfe", "pktype",
+     stream << indent() << "std::time_t curr_time = std::time(nullptr);"
+            << endl;
+     stream << indent() << "char curr_time_str["
+            << "sizeof(\"yyyy-mm-dd, hh:mm:ss\")];" << endl;
+     stream << indent() << "std::strftime(curr_time_str, sizeof(curr_time_str),"
+            << " \"%F, %T\", std::localtime(&curr_time));" << endl;
+     stream << indent() << "std::cerr << \"[\" << curr_time_str << \"] "
+            << "Error trying to archive cyk loop progress.\\n\";" << endl;
+     stream << indent() << "boost::filesystem::remove(tmp_out_cyk_path);"
+            << endl;
+     dec_indent();
+     stream << indent() << "}" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl << endl;
+     dec_indent();
+  }
+
+  void load_cyk_indices(Printer::Base &stream, size_t n_tracks) {
+     inc_indent();
+     stream << indent() << "void load_cyk_indices() {" << endl;
+     inc_indent();
+     stream << indent() << "// read the cyk loop indices from disk " << endl;
+     stream << indent() << "try {" << endl;
+     inc_indent();
+     stream << indent() << "std::ifstream array_fin(in_archive_path.c_str(), "
+            << "std::ios::binary);" << endl;
+     stream << indent() << "if (!(array_fin.good())) {" << endl;
+     stream << indent() << "  throw std::ifstream::failure(\"\");" << endl;
+     stream << indent() << "}" << endl;
+     stream << indent() << "boost::archive::binary_iarchive "
+                            "array_in(array_fin);" << endl;
+     for (size_t i = 0; i < n_tracks; i++) {
+       stream << indent() << "array_in >> t_" << i << "_i;" << endl;
+       stream << indent() << "array_in >> t_" << i << "_j;" << endl;
+     }
+     stream << indent() << "array_fin.close();" << endl << endl;
+     stream << indent() << "std::cerr << \"Info: Successfully loaded "
+            << "cyk loop indices. \"" << endl;
+     stream << indent() << "          << \"Will continue calculating from here."
+            << "\" << std::endl;" << endl;
+     dec_indent();
+     stream << indent() << "} catch (const std::ifstream::failure &e) {"
+            << endl;
+     inc_indent();
+     stream << indent() << "std::cerr << \"Error: \\\"\" + "
+            << "in_archive_path.string() + \"\\\" "
+            << "archive\"" << endl;
+     stream << indent() << "          << " << "\" could not be found.\\n\";"
+            << endl;
+     stream << indent() << "std::exit(1);" << endl;
+     dec_indent();
+     stream << indent() << "} catch (const std::exception &e) {" << endl;
+     inc_indent();
+     stream << indent() << "std::cerr << \"Error \\\"\" << e.what() << \"\\\" "
+                            "trying to read cyk loop indices!\\n\";" << endl;
+     stream << indent() << "std::exit(1);" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl;
+     dec_indent();
+     stream << indent() << "}" << endl << endl;
+     dec_indent();
+  }
+
+  void map_1d_to_2d(Printer::Base &stream) {
```

Comment:
when is this fct called?

Reply:
This function will be called in the `restore_string_links` function if the table contains `String` types and we checkpoint in `cyk` mode. Since there is no `tabulated` vector in `cyk` mode, I can't just mark the broken Lists/Strings as not tabulated to force the DP algorithm to recalculate them. Instead I have to manually call the appropriate non-terminal functions to recalculate the broken Lists/Strings. These non-terminal functions take two indices as input though while I only have access to the index of the broken List in its `std::vector`. So I need to map the 1D vector index to a `row` and `column` index as if the 1D vector were a 2D (upper triangular) matrix, which this function does.  

A little complicated, I know.

Reply:
This function is now no longer needed (see last comment).